scipy/linalg/decomp_svd.py

```diff
@@ -279,6 +286,102 @@ def diagsvd(s, M, N):
         raise ValueError("Length of s must be M or N.")
 
 
+def higher_order_svd(
+    a: npt.ArrayLike, *, full_tensor: bool = True, check_finite: bool = True
+) -> Tuple[List[numpy.ndarray], numpy.ndarray]:
+    """Higher-order SVD (HOSVD)
+
+    Factorizes the M-D tensor `a` into a list of M matrices ``U_k`` containing
+    the left singular vectors of the unfolded tensor with respect to the
+    ``k``\\ th axis and a core tensor ``S`` such that
+    :math:`a = (U_1, U_2, \\dots, U_M) \\cdot S`.
+
+    Parameters
+    ----------
+    a : array_like
+        Tensor to decompose
+    full_tensor : bool, optional
+        If True (default), ``S`` is of the same shape as the input tensor `a`
+        and all ``U_k`` are of the shape ``(n, n)`` with ``n`` corresponding to
+        the ``k``\\ th axis of ``S``.
+        If False, the shapes of the ``U_k`` are ``(n, m)`` with ``n``
+        corresponding to the size of the ``k``\\ th axis of the input `a` and
+        ``m`` to the maybe reduced size of the ``k``\\ th axis of the core
+        tensor.
+    check_finite : bool, optional
+        Whether to check that the input matrix contains only finite numbers.
+        Disabling may give a performance gain, but may result in problems
+        (crashes, non-termination) if the inputs do contain infinities or NaNs.
+
+    Returns
+    -------
+    U : list of ndarray
+        List of the unitary matrices ``U_k`` with left singular vectors of the
+        tensor unfolding corresponding to the ``k``\\ th axis.
+        Of shape ``(M, M)`` or ``(M, K)``, depending on `full_tensor`.
+    S : ndarray
+        The core tensor.
+        Of same or reduced shape compared to the input `a`, depending on
+        `full_tensor`.
+
+    Raises
+    ------
+    LinAlgError
+        If SVD computation does not converge.
+
+    .. versionadded:: 1.8.0
+
+    See Also
+    --------
+    svd : Singular value decomposition of a matrix
+
+    References
+    ----------
+    .. [1] Lieven De Lathauwer, Bart De Moor, and Joos Vandewalle (2000)
+           "A Multilinear Singular Value Decomposition"
+           SIAM Journal on Matrix Analysis and Applications,
+           21 (4), pp. 1253-1278. ISSN 1095-7162
+           https://doi.org/10.1137/S0895479896305696
+
+    Examples
+    --------
+    >>> from numpy.random import default_rng
+    >>> from scipy import linalg
+    >>> rng = default_rng()
```

Comment:
I guess for this test the numerical values don't matter. We generally write it as `np.random.default_rng()` so that in the background it gets a seed (and NumPy is imported in the background to and not needed `np`).

Reply:
Thanks for the review. I have changed this, too.